### PR TITLE
Custom destinations for modules

### PR DESCRIPTION
**Pull request by ivan-kostko**
This feature enables module management with specific destinations, while back compatibility is intact.

Centrally managed: `Terrafile` is located in "root" directory of your terraform code, managing modules in all subfolders / stacks
An example of using `Terrafile` in a root directory:

Let's assume following directory structure structure

```
.
├── iam
│   ├── main.tf
│   └── .....tf
├── networking
│   ├── main.tf
│   └── .....tf
├── onboarding
.
.
.
├── some-other-stack
└── Terrafile
```

In above scenarion Terrafile is not in every single folder but in the "root" of terraform code.

An example usage of centrally managed modules:

```
tf-aws-vpc:
    source:  "git@github.com:terraform-aws-modules/terraform-aws-vpc"
    version: "v1.46.0"
    destination:
        - networking
tf-aws-iam:
    source:  "git@github.com:terraform-aws-modules/terraform-aws-iam"
    version: "v5.11.1"
    destination:
        - iam
tf-aws-s3-bucket:
    source:  "git@github.com:terraform-aws-modules/terraform-aws-s3-bucket"
    version: "v3.6.1"
    destination:
        - networking
        - onboarding
        - some-other-stack
```

The `destination` of module is an array of directories (stacks) where the module should be used.
The module itself is fetched once and copied over to designated destionations.
Final destination of the module is handled in a similar way as in first approach: `$destination/$module_path/$module_key`.

The output of the run is exactly the same in both options.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,7 +21,11 @@ curl -L https://github.com/coretech/terrafile/releases/download/v{VERSION}/terra
 ## How to use
 Terrafile expects a file named `Terrafile` which will contain your terraform module dependencies in a yaml like format.
 
-An example Terrafile:
+There are two approaches that can be used for managing your modules depending on the structure of your terraform code:
+1. The default approach: `Terrafile` is located directly in the directory where terraform is run
+2. Centrally managed: `Terrafile` is located in "root" directory of your terraform code, managing modules in all subfolders / stacks
+
+An example of default approach (#1) to `Terrafile`:
```
Comment (nritholtz) [inline README.md:28]:
```suggestion
### Default Approach
An example of default approach (#1) to `Terrafile`
```

File: README.md [hunk 2]
```diff
@@ -34,24 +38,74 @@ tf-aws-vpc-experimental:
...
 Terraform modules exported to custom directory
 ```sh
 $ terrafile -p custom_directory
-INFO[0000] [*] Checking out master of git@github.com:terraform-aws-modules/terraform-aws-vpc  
-INFO[0001] [*] Checking out v1.46.0 of git@github.com:terraform-aws-modules/terraform-aws-vpc  
+INFO[0000] [*] Checking out master of git@github.com:terraform-aws-modules/terraform-aws-vpc
+INFO[0001] [*] Checking out v1.46.0 of git@github.com:terraform-aws-modules/terraform-aws-vpc
+```
+
+An example of using `Terrafile` in a root directory:
```
Comment (nritholtz) [inline README.md:59]:
```suggestion
### Centrally Managed Approach
An example of using `Terrafile` in a root directory (#2):
```

**Review by nritholtz — CHANGES_REQUESTED** [review]
Added some typo fixes and minor suggestions on reworking the implementation.

File: .gitignore [hunk 3]
```diff
@@ -17,4 +17,5 @@ terrafile
 
 #IDEs
 .idea/
-dist/
\ No newline at end of file
+dist/
+.DS_Store
```
[the remark this answers is not in the document]
Comment (ivan-kostko) [reply .gitignore:22]: Done

File: main.go [hunk 4]
```diff
@@ -31,12 +30,13 @@ import (
 )
 
 type module struct {
-	Source  string `yaml:"source"`
-	Version string `yaml:"version"`
+	Source      string   `yaml:"source"`
+	Version     string   `yaml:"version"`
+	Destination []string `yaml:"destination"`
```
[the remark this answers is not in the document]
Comment (ivan-kostko) [reply main.go]: Accepted and adapted

File: main.go [hunk 5]
```diff
@@ -69,31 +71,70 @@ func main() {
...
+				go func(dst string, m module, key string) {
+					defer wg.Done()
+					if err := os.MkdirAll(dst, os.ModePerm); err != nil {
+						log.Errorf("failed to create folder %s due to error: %s", dst, err)
+						return
+					}
+					// Delete previously copied artifacts
+					os.RemoveAll(filepath.Join(dst, key))
+					moduleSrc := filepath.Join(firstDestination, key)
+					cmd := exec.Command("cp", "-Rf", moduleSrc, dst)
```
[the remark this answers is not in the document]
Comment (ivan-kostko) [reply main.go]: Patrially Accepted and adapted. Hardlinks won't work for folders. However, SymLink does the job ;-). Please, have a look on latest changes

File: main.go [hunk 6]
```diff
@@ -53,47 +53,105 @@ func init() {
...
 	for key, mod := range config {
 		wg.Add(1)
 		go func(m module, key string) {
 			defer wg.Done()
-			gitClone(m.Source, m.Version, key)
-			_ = os.RemoveAll(filepath.Join(opts.ModulePath, key, ".git"))
+
+			// path to clone module
+			cloneDestination := opts.ModulePath
+			// list of paths to link module to. empty, uless Destinations are more than 1 location
```
Comment (nritholtz) [inline main.go:108]:
```suggestion
			// list of paths to link module to. empty, unless Destinations are more than 1 location
```

File: main_test.go [hunk 7]
```diff
@@ -91,10 +127,55 @@ func TestTerraformWithTerrafilePath(t *testing.T) {
...
+		"testdata/some-other-stack/vendor/modules": {
+			"tf-aws-s3-bucket": {
+				"repository": "git@github.com:terraform-aws-modules/terraform-aws-s3-bucket",
+				"version":    "v3.6.1",
+			},
+		},
+	} {
+		for moduleName, cloneOptions := range checkout {
+			testModuleLocation := path.Join(workingDirectory, dst, moduleName+"__test")
+			os.RemoveAll(testModuleLocation)
```
Comment (nritholtz) [inline main_test.go:164]:
```suggestion
			_ = os.RemoveAll(testModuleLocation)
```

**Review by nritholtz — CHANGES_REQUESTED** [review]
awesome work! two very minor fixes, otherwise looks good to me

**Comment by ivan-kostko** [comment]
Hi @nritholtz ,

Sorry for off topic, but as we discussed inside our team, the tool is missing `-clean` flag. The issue appears whenever some module is removed from file, but tool does not remove it from filesystem. Hence, we will prepare a separate PR introducing this functionality as soon as this PR is merged.

**Review by nritholtz — APPROVED** [review]

**Pull request merged**